server/src/main/java/org/opensearch/search/aggregations/bucket/histogram/DateHistogramAggregator.java

```diff
@@ -117,34 +119,40 @@ class DateHistogramAggregator extends BucketsAggregator implements SizedBucketAg
 
         bucketOrds = LongKeyedBucketOrds.build(context.bigArrays(), cardinality);
 
-        fastFilterContext = new FastFilterRewriteHelper.FastFilterContext(
-            context,
-            new DateHistogramAggregationType(
-                valuesSourceConfig.fieldType(),
-                valuesSourceConfig.missing() != null,
-                valuesSourceConfig.script() != null,
-                hardBounds
-            )
-        );
-        if (fastFilterContext.isRewriteable(parent, subAggregators.length)) {
-            fastFilterContext.buildRanges(Objects.requireNonNull(valuesSourceConfig.fieldType()));
-        }
-    }
+        optimizationContext = new OptimizationContext(new DateHistogramAggregatorBridge() {
+            @Override
+            public boolean canOptimize() {
+                return canOptimize(valuesSourceConfig);
+            }
 
-    private class DateHistogramAggregationType extends FastFilterRewriteHelper.AbstractDateHistogramAggregationType {
+            @Override
+            public void prepare() throws IOException {
+                buildRanges(context);
+            }
 
-        public DateHistogramAggregationType(MappedFieldType fieldType, boolean missing, boolean hasScript, LongBounds hardBounds) {
-            super(fieldType, missing, hasScript, hardBounds);
-        }
+            @Override
+            protected Rounding getRounding(long low, long high) {
+                return rounding;
+            }
 
-        @Override
-        protected Rounding getRounding(long low, long high) {
-            return rounding;
-        }
+            @Override
+            protected Rounding.Prepared getRoundingPrepared() {
+                return preparedRounding;
+            }
+
+            @Override
+            protected long[] processHardBounds(long[] bounds) {
+                return super.processHardBounds(bounds, hardBounds);
+            }
 
-        @Override
-        protected Rounding.Prepared getRoundingPrepared() {
-            return preparedRounding;
+            @Override
+            protected Function<Long, Long> bucketOrdProducer() {
+                return (key) -> bucketOrds.add(0, preparedRounding.round((long) key));
+            }
+
+        });
+        if (optimizationContext.canOptimize(parent, subAggregators.length, context)) {
```

Comment:
I think we can clean this up a bit by moving all of this initialization logic inside of the OptimizationContext ctor. ie. `        optimizationContext = new OptimizationContext(aggregatorBridge, parent, subAggregators.length, context);`

The ctor would be something like:
```java
    public OptimizationContext(AggregatorBridge aggregatorBridge, final Object parent, final int subAggLength, SearchContext context) {
        this.aggregatorBridge = aggregatorBridge;
        this.canOptimize = canOptimize(parent, subAggLength, context);
    }
```

This has the added benefit of canOptimize being immutable + reduces the chattiness going between these components.

Reply:
This makes the change look simpler in aggregator, so adopted